lib/sip2/client.rb

```diff
@@ -5,18 +5,33 @@ module Sip2
   # Sip2 Client
   #
   class Client
-    def initialize(host:, port:, ignore_error_detection: false, timeout: nil)
+    def initialize(host:, port:, ignore_error_detection: false, timeout: nil, ssl_context: nil)
       @host = host
       @port = port
       @ignore_error_detection = ignore_error_detection
       @timeout = timeout || NonBlockingSocket::DEFAULT_TIMEOUT
+      @ssl_context = ssl_context
     end
 
+    # rubocop:disable Metrics/MethodLength
     def connect
-      socket = NonBlockingSocket.connect @host, @port, @timeout
-      yield Connection.new(socket, @ignore_error_detection) if block_given?
+      socket = NonBlockingSocket.connect host: @host, port: @port, timeout: @timeout
+
+      # If we've been provided with an SSL context then use it to wrap out existing connection
+      if @ssl_context
+        socket = ::OpenSSL::SSL::SSLSocket.new socket, @ssl_context
+        socket.hostname = @host # Needed for SNI
+        socket.sync_close = true
+        socket.connect
+        socket.post_connection_check @host # Validate the peer certificate matches the host
+      end
```

Comment:
Is any of this needed when not `block_given?`? It looks like we could include all this in the `if block_given?` below.

Reply:
yes, we do need it. If a block isn't given then the code is simply going to check that it can connect. If someone specifies the `ssl_context` and calls to `connect` we need to actually use it and connect the SSL socket